test/loadEnvironment.test.ts

```diff
@@ -0,0 +1,104 @@
+import * as fs from "fs"
+import { loadEnvironment } from "../src/loadEnvironment"
+
+const getContentfulEnvironmentFileFactory = jest.fn((_type: string) => {
+  return () => ({
+    getContentTypes: () => [],
+    getLocales: () => [],
+  })
+})
+
+jest.mock(
+  require("path").resolve(process.cwd(), "./getContentfulEnvironment.js"),
+  () => getContentfulEnvironmentFileFactory("js"),
+  { virtual: true },
+)
+
+jest.mock(
+  require("path").resolve(process.cwd(), "./getContentfulEnvironment.ts"),
+  () => getContentfulEnvironmentFileFactory("ts"),
+  { virtual: true },
+)
+
+const tsNodeRegistererEnabled = jest.fn()
+const tsNodeRegister = jest.fn()
+
+jest.mock("ts-node", () => ({ register: tsNodeRegister }))
+
+describe("loadEnvironment", () => {
+  beforeEach(() => {
+    jest.resetAllMocks()
+    jest.restoreAllMocks()
+
+    getContentfulEnvironmentFileFactory.mockReturnValue(() => ({
+      getContentTypes: () => [],
+      getLocales: () => [],
+    }))
+    tsNodeRegister.mockReturnValue({ enabled: tsNodeRegistererEnabled })
+  })
+
+  describe("when getContentfulEnvironment.ts exists", () => {
```

Comment:
I love the test coverage here! I'm only missing the positive test case here:  when .ts exists it calls the factory with `ts` and not `js`

Reply:
The reason I don't have that case is because it actually just flat out fails due to issues with the mocks - having an inverse of the "it requires the javascript config" fails so long as the other `describe` tests exist (i.e. sticking `.only` on "it requires the typescript config" and "it disables the registerer afterwards" results it in passing), and if you add the specific `expect`s to "when called multiple times -> it re-uses the registerer" that passes fine; alternatively disabling resetting/restoring mocks gets the test passing, but introduces potential leakage across the tests and the rest start failing.

I actually think that this could be (at least in part) due to a bug within Jest, since you're more than a couple of major versions out of date (currently v26, latest is v29), but overall I'm confident between our manual testing and the existing test suites that we have sufficient coverage (which is 100%) without this test so would prefer to omit it for now rather than keep sinking a heavy amount of time into it.

I'm happy to follow this PR up with some general updates and also tackle a few open issues, which could give an opportunity to look back into this :)

Reply:
(alternatively I can add the `expect`s to the aforementioned "it re-uses the registerer" test if you like - it'll be a bit odd, but at least then we'll be covering that path explicitly)

Reply:
I agree it's weird. Seems like a bug in Jest or something we're not seeing.

I tried using `jest.resetModules()`, which should clean up the state between tests, but that only works if the tests are ran in a specific order: adding the test just before the `disables the registerer afterwards` test case.

```ts
    })

    it("requires the typescript config", async () => {
      await loadEnvironment()

      expect(getContentfulEnvironmentFileFactory).toHaveBeenCalledWith("ts")
      expect(getContentfulEnvironmentFileFactory).not.toHaveBeenCalledWith("js")
    })

    it("disables the registerer afterwards", async () => {
```

 If that sounds good to you, let's do that for know, even if it's not ideal, and we can get this one merged 🚀

Reply:
🤦 duh `jest.resetModules` is exactly what we need here, because it's a `require` - thanks, I'll push that up